lib/libvfio-user.c

```diff
@@ -524,25 +524,57 @@ handle_dma_unmap(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg,
                  struct vfio_user_dma_region *region)
 {
     int ret;
+    char *bitmap;
     char rstr[1024];
 
     assert(vfu_ctx != NULL);
     assert(msg != NULL);
     assert(region != NULL);
 
-    if (msg->in_size != sizeof(*region)) {
+    if (msg->in_size < sizeof(*region)) {
         vfu_log(vfu_ctx, LOG_ERR, "bad size of DMA unmap region %zu",
                 msg->in_size);
         return ERROR_INT(EINVAL);
     }
 
-    snprintf(rstr, sizeof(rstr), "[%#lx, %#lx) offset=%#lx "
-            "prot=%#x flags=%#x", region->addr, region->addr + region->size,
-            region->offset, region->prot, region->flags);
+    snprintf(rstr, sizeof(rstr), "[%#lx, %#lx) offset=%#lx prot=%#x flags=%#x",
+             region->addr, region->addr + region->size,
+             region->offset, region->prot, region->flags);
 
     vfu_log(vfu_ctx, LOG_DEBUG, "removing DMA region %s", rstr);
 
-    if (region->flags != 0) {
+    if (region->flags == VFIO_DMA_UNMAP_FLAG_GET_DIRTY_BITMAP) {
+        if (msg->in_size != sizeof(struct vfio_user_dma_region) + sizeof(struct vfio_user_bitmap)) {
+            vfu_log(vfu_ctx, LOG_ERR, "bad message size %#lx", msg->in_size);
+            return ERROR_INT(EINVAL);
+        }
+        /*
+         * TODO this could be a separate function, but the implementation is
+         * temporary anyway since we're moving dirty page tracking out of
+         * the DMA controller.
+         */
+        ret = dma_controller_dirty_page_get(vfu_ctx->dma,
+                                            (vfu_dma_addr_t)region->addr,
+                                            region->size,
+                                            region->bitmap->pgsize,
+                                            region->bitmap->size,
+                                            &bitmap);
+        if (ret < 0) {
+            vfu_log(vfu_ctx, LOG_ERR, "failed to get dirty page bitmap: %m");
+            return -1;
+        }
+        msg->out_data = malloc(region->bitmap->size);
```

Comment:
I thought you were going to do:

```
msg->out_data = region->bitmap;
msg->out_size = region->bitmap->size;
region->bitmap = NULL;
```

??

Reply:
I thought I was, but I was wrong in my example: `region->bitmap` is the struct in the request where the size/pgsize is set, not the actual dirty page bitmap. The dirty page bitmap lives in `dma_memory_region_t` and I'm not sure it's worth the effort to expose this as we're removing dirty page logging from the DMA controller in the mid-term. So I opted to allocate memory as that seems the easiest thing to do for now.